src/poller.rs

```diff
@@ -1,38 +1,91 @@
-use std::path::Path;
+use std::fs;
+use std::path::{Path, PathBuf};
 use std::process;
 use std::time::Instant;
 
 use tokio::time;
 use tracing::{error, info};
 
-use crate::config::Config;
+use crate::config::{Config, WatchConfig};
 use crate::log::Operation;
 use crate::snapshots;
 
+/// Checks the provided `child_path` is a directory.
+/// If either `includes` or `excludes` are set,
+/// checks whether the path is included/excluded respectively.
+fn is_valid_directory(base_path: &Path, child_path: &PathBuf, value: &WatchConfig) -> bool {
+    if !child_path.is_dir() {
+        return false;
+    }
+
+    let includes = &value.include;
+    let excludes = &value.exclude;
+
+    if includes.len() > 0 {
+        let relative = child_path.strip_prefix(base_path).unwrap().to_str().unwrap();
+        includes
+            .iter()
+            .find(|&include| include.starts_with(relative))
+            .is_some()
+    } else if excludes.len() > 0 {
+        let relative = child_path.strip_prefix(base_path).unwrap().to_str().unwrap();
+        excludes
+            .iter()
+            .find(|&exclude| exclude.starts_with(relative))
+            .is_none()
+    } else {
+        true
+    }
+}
+
+/// If the directory is a repo, attempts to create a snapshot.
+/// Otherwise, recurses into each child directory.
 #[tracing::instrument]
-fn process_directory(path: &Path) {
-    let mut op: Option<snapshots::CaptureStatus> = None;
-    let mut error: Option<String> = None;
-    let start_time = Instant::now();
+fn process_directory(base_path: &Path, current_path: &Path, value: &WatchConfig, depth: u8) {
+    if snapshots::is_repo(current_path) {
+        let mut op: Option<snapshots::CaptureStatus> = None;
+        let mut error: Option<String> = None;
+        let start_time = Instant::now();
 
-    match snapshots::capture(path) {
-        Ok(Some(status)) => op = Some(status),
-        Ok(None) => (),
-        Err(err) => {
-            error = Some(format!("{}", err));
+        match snapshots::capture(current_path) {
+            Ok(Some(status)) => op = Some(status),
+            Ok(None) => (),
+            Err(err) => {
+                error = Some(format!("{}", err));
+            }
         }
-    }
 
-    let latency = (Instant::now() - start_time).as_secs_f32();
-    let repo = path.to_str().unwrap_or("<invalid path>").to_string();
-    let operation = Operation::Snapshot {
-        repo,
-        op,
-        error,
-        latency,
-    };
-    if operation.should_log() {
-        info!(operation = %serde_json::to_string(&operation).unwrap(),"info_operation")
+        let latency = (Instant::now() - start_time).as_secs_f32();
+        let repo = current_path
+            .to_str()
```

Comment:
A better rustacean might have input here (maybe `cargo clippy`?) but I think there might be a better way to do this sequence of calls.

Reply:
Same as above - `format` just moved this about. `clippy` isn't saying anything about it. Not really sure how to improve it myself I'm afraid.

Reply:
I highly value your contributions and I'll gladly merge it as long as 

1. there's no regressions
2. it doesn't force us into a tough situation regarding backward compatibility

most of my comments are intended to be fixed only if you want to, and I would merge before you even got a chance to fix them, if I had the opportunity. There's nothing wrong with extra PRs for superficial stuff.